stock_product_variant_mto/models/product_template.py

```diff
@@ -1,24 +1,42 @@
 # Copyright 2023 Camptocamp SA
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
 
-from odoo import models, fields
+from odoo import _, api, models, fields
 
 class ProductTemplate(models.Model):
     _inherit = "product.template"
 
-    has_mto_route_changed = fields.Boolean()
-
     def write(self, values):
         if not "route_ids" in values:
             return super().write(values)
-        # This route_ids change will trigger variant's _compute_is_mto.
-        # We want to set variant's is_mto to True only if their
-        # template has been set to True here ↓
+        # As _compute_is_mto cannot use api.depends (or it would reset MTO
+        # route on variants as soon as there is a change on the template routes),
+        # we need to check which template in self had MTO route activated
+        # or deactivated to force the recomputation of is_mto on variants
         mto_route = self.env.ref("stock.route_warehouse0_mto")
         template_not_mto_before = self.filtered(lambda t: mto_route not in t.route_ids)
         res = super().write(values)
-        template_mto_after = self.filtered(lambda t: mto_route in t.route_ids)
-        # Templates where mto route has changed are those where
-        # the mto route has been set
-        templates_mto_set = template_not_mto_before & template_mto_after
-        templates_mto_set.has_mto_route_changed = True
+        templates_mto_after = self.filtered(lambda t: mto_route in t.route_ids)
+        templates_mto_added = template_not_mto_before & templates_mto_after
+        templates_mto_removed = (self - template_not_mto_before) & (self - templates_mto_after)
+        (templates_mto_added | templates_mto_removed).product_variant_ids._compute_is_mto()
+
+    @api.onchange("route_ids")
+    def onchange_route_ids(self):
+        mto_route = self.env.ref("stock.route_warehouse0_mto", raise_if_not_found=False)
+        if mto_route not in self._origin.route_ids and mto_route in self.route_ids._origin:
```

Comment:
I guess we can make the condition more selective and raise the warning only if we have multiple variants, and these variants have different MTO settings.